p2panda-net/src/sync/tests.rs

```diff
@@ -0,0 +1,456 @@
+        rx_acceptor.recv().await,
+        Some(ToEngineActor::SyncFailed { .. })
+    ));
+
+    // Expected handler results.
+    assert_eq!(
+        initiator_handle.await.unwrap(),
+        Err(SyncError::Critical(
+            "something really bad happened in the initiator".into(),
+        ))
+    );
+    assert_eq!(
+        acceptor_handle.await.unwrap(),
+        // The acceptor failed as well, but only with an "unexpected behaviour" error over the
+        // unexpectedly closed pipe.
+        Err(SyncError::UnexpectedBehaviour("broken pipe".into()))
+    );
+}
+
+#[tokio::test]
+async fn initiator_sends_topic_twice() {
+    let (mut rx_initiator, mut rx_acceptor, initiator_handle, acceptor_handle) =
+        run_sync_impl(FailingProtocol::InitiatorSendsTopicTwice).await;
+
+    // Expected initiator messages.
+    assert!(matches!(
+        rx_initiator.recv().await,
+        Some(ToEngineActor::SyncStart { .. })
+    ));
+
+    assert!(matches!(
+        rx_initiator.recv().await,
+        Some(ToEngineActor::SyncHandshakeSuccess { .. })
+    ));
+
+    assert!(matches!(
+        rx_initiator.recv().await,
+        Some(ToEngineActor::SyncDone { .. })
+    ));
+
+    // Expected acceptor messages.
+    assert!(matches!(
+        rx_acceptor.recv().await,
+        Some(ToEngineActor::SyncStart { .. })
+    ));
+
+    assert!(matches!(
+        rx_acceptor.recv().await,
+        Some(ToEngineActor::SyncFailed { .. })
+    ));
+
+    assert_eq!(initiator_handle.await.unwrap(), Ok(()));
+    assert_eq!(
+        acceptor_handle.await.unwrap(),
+        // This is _not_ a critical error as the acceptor protocol implementation handled the
+        // protocol violation (sending topic twice) by itself.
+        Err(SyncError::UnexpectedBehaviour(
+            "received topic too often".into(),
+        ))
+    );
+}
+
+#[tokio::test]
+async fn acceptor_fails_critical() {
+    let (mut rx_initiator, mut rx_acceptor, initiator_handle, acceptor_handle) =
+        run_sync_impl(FailingProtocol::AcceptorFailsCritical).await;
+
+    // Expected initiator messages.
+    assert!(matches!(
+        rx_initiator.recv().await,
+        Some(ToEngineActor::SyncStart { .. })
+    ));
+
+    assert!(matches!(
+        rx_initiator.recv().await,
+        Some(ToEngineActor::SyncHandshakeSuccess { .. })
+    ));
+
+    assert!(matches!(
+        rx_initiator.recv().await,
+        // Initiator can end the session without any issues even when the acceptor fails.
+        // @TODO: Do we want to detect the closed connection from the remote peer?
+        Some(ToEngineActor::SyncDone { .. })
+    ));
+
+    // Expected acceptor messages.
+    assert!(matches!(
+        rx_acceptor.recv().await,
+        Some(ToEngineActor::SyncStart { .. })
+    ));
+
+    assert!(matches!(
+        rx_acceptor.recv().await,
+        Some(ToEngineActor::SyncFailed { .. })
+    ));
+
+    // Expected handler results.
+    assert_eq!(initiator_handle.await.unwrap(), Ok(()));
```

Comment:
Interestingly we can't detect a closed pipe if the acceptor "crashed" .. I'm not sure why that is.